### PR TITLE
Check API permissions

### PR DESCRIPTION
**Pull request by ortyomka**

### CODE REVIEW

File: core/api/rpc/make.hpp [hunk 1]
```diff
@@ -42,7 +48,14 @@ namespace fc::api {
         [&](auto &jparams,
             rpc::Respond respond,
             rpc::MakeChan make_chan,
-            rpc::Send send) {
+            rpc::Send send,
+            const rpc::Perms &perms) {
+          if (std::find(perms.begin(), perms.end(), method.getPerm())
+              == perms.end()) {
```
Comment (turuslan) [inline core/api/rpc/make.hpp:54]: this implies that `generate(admin)` wont work and should be `generate(admin, read, write)`?

File: core/miner/main/main.cpp [hunk 2]
```diff
@@ -441,14 +446,19 @@ namespace fc {
...
 
-    OUTCOME_TRY(api_secret, loadApiSecret(config.join("jwt_secret")));
     api::fillAuthApi(mapi, api_secret, log());
 
     std::map<std::string, std::shared_ptr<api::Rpc>> mrpc;
-    mrpc.emplace("/rpc/v0", api::makeRpc(*mapi));
+    mrpc.emplace(
+        "/rpc/v0",
+        api::makeRpc(*mapi,
+                     std::bind(mapi->AuthVerify, std::placeholders::_1)));
```
Comment (turuslan) [inline core/miner/main/main.cpp:455]: maybe should split `verifyAuthToken(secret, token)` like `generateAuthToken` and use lambda

File: core/sector_storage/impl/fetch_handler.cpp [hunk 3]
```diff
@@ -166,36 +162,44 @@ namespace fc::sector_storage {
...
       const std::shared_ptr<stores::LocalStore> &local_store) {
     return [local = local_store, logger = server_logger](
-               const http::request<http::dynamic_body> &request)
-               -> api::WrapperResponse {
-      // TODO(artyom-yurin): [FIL-259] Check token
+               const http::request<http::dynamic_body> &request,
+               const api::Perms &perms) -> api::WrapperResponse {
+      if (std::find(
+              perms.begin(), perms.end(), primitives::jwt::kAdminPermission)
+          == perms.end()) {
```
Comment (turuslan) [inline core/sector_storage/impl/fetch_handler.cpp:200]: maybe should reuse this check with api

File: core/api/rpc/ws.cpp [hunk 4]
```diff
@@ -140,9 +144,39 @@ namespace fc::api {
...
+    auto it = request.find(http::field::authorization);
+    if (it != request.cend()) {
+      auto auth_token = (it->value());
+
+      if (auth_token.empty()) {
+        return std::string();
+      }
+
+      const std::string perfix = "Bearer ";
+      if (not auth_token.starts_with(perfix)) {
```
Comment (turuslan) [inline core/api/rpc/ws.cpp:162]: [c++20](https://en.cppreference.com/w/cpp/string/basic_string/starts_with), maybe replace with [boost](https://www.boost.org/doc/libs/1_72_0/doc/html/boost/algorithm/starts_with.html)

File: core/api/rpc/make.hpp (hunk 1 above)
Comment (ortyomka) [reply core/api/rpc/make.hpp:54]: admin requires only for /remote

**Review by turuslan — APPROVED** [review]

File: core/api/rpc/ws.cpp [hunk 5]
```diff
@@ -140,9 +144,39 @@ namespace fc::api {
...
     beast::flat_buffer buffer;
+    Perms perms;
     Rpc rpc;
   };
 
+  std::optional<std::string> getToken(
+      const http::request<http::dynamic_body> &request) {
+    auto it = request.find(http::field::authorization);
+    if (it != request.cend()) {
+      auto auth_token = (it->value());
```
Comment (Alexey-N-Chernyshov) [inline core/api/rpc/ws.cpp:155]:
```suggestion
      const auto auth_token = (it->value());
```

File: core/api/rpc/ws.cpp [hunk 6]
```diff
@@ -311,6 +360,32 @@ namespace fc::api {
...
   }
 
+  RouteHandler makeAuthRoute(AuthRouteHandler &&handler,
+                             rpc::AuthFunction &&auth) {
+    return [handler{std::move(handler)}, auth{std::move(auth)}](
+               const http::request<http::dynamic_body> &request)
+               -> WrapperResponse {
+      Perms perms = kDefaultPermission;
+      if (auth) {
+        auto maybe_token = getToken(request);
```
Comment (Alexey-N-Chernyshov) [inline core/api/rpc/ws.cpp:370]:
```suggestion
        const auto maybe_token = getToken(request);
```

File: core/api/rpc/ws.cpp [hunk 7]
```diff
@@ -311,6 +360,32 @@ namespace fc::api {
...
+        if (not maybe_token.has_value()) {
+          return makeErrorResponse(request, http::status::unauthorized);
+        }
+
+        if (not maybe_token.value().empty()) {
+          auto maybe_perms =
```
Comment (Alexey-N-Chernyshov) [inline core/api/rpc/ws.cpp:376]:
```suggestion
          const auto maybe_perms =
```

File: core/api/rpc/rpc.hpp [hunk 8]
```diff
@@ -42,21 +43,41 @@ namespace fc::api {
...
+  using primitives::jwt::kDefaultPermission;
+  using primitives::jwt::Permission;
   using rapidjson::Value;
 
   using OkCb = std::function<void(bool)>;
   using Respond =
       std::function<void(boost::variant<Response::Error, Document>)>;
   using Send = std::function<void(std::string, Document, OkCb)>;
   using MakeChan = std::function<uint64_t()>;
+  using Perms = std::vector<Permission>;
```
Comment (wer1st) [inline core/api/rpc/rpc.hpp:55]: I suggest to use `Permissions` instead of `Perms`, it looks better

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: core/api/rpc/ws.cpp [hunk 9]
```diff
@@ -181,8 +215,23 @@ namespace fc::api {
         for (const auto &api : rpc) {
           // API version is specified in request (e.g. '/rpc/v0')
           if (request.target().starts_with(api.first)) {
+            auto maybe_token = getToken(request);
```
Comment (wer1st) [inline core/api/rpc/ws.cpp:218]:
```suggestion
            const auto maybe_token = getToken(request);
```

File: core/api/rpc/ws.cpp [hunk 10]
```diff
@@ -181,8 +215,23 @@ namespace fc::api {
...
+            if (not maybe_token.has_value()) {
+              w_response =
+                  makeErrorResponse(request, http::status::unauthorized);
+              return doWrite();
+            }
+
+            auto maybe_perms = api.second->getPerms(maybe_token.value());
```
Comment (wer1st) [inline core/api/rpc/ws.cpp:225]:
```suggestion
           const  auto maybe_perms = api.second->getPerms(maybe_token.value());
```

File: core/api/rpc/ws.cpp (hunk 7 above)
Comment (ortyomka) [reply core/api/rpc/ws.cpp:376]: It would moved

File: core/api/rpc/ws.cpp (hunk 10 above)
Comment (ortyomka) [reply core/api/rpc/ws.cpp:225]: It would moved

**Review by wer1st — APPROVED** [review]

**Pull request merged**